### PR TITLE
multihash-and-fixes

### PR DESCRIPTION
**Pull request by art-gor**
1) Multihash has immutable shared_ptr<const Data> inside, including precalculated std::hash, allows for performance while copying, sharing, using as key in containers
2) UVarint stuff in Multihash creation changed to impl with no dynamic allocations
3) Kademlia's PeerRoutingTable fixed (bugs were detected while applying p.1)

### CODE REVIEW

File: include/libp2p/protocol/kademlia/impl/peer_routing_table_impl.hpp [hunk 1]
```diff
@@ -50,39 +51,89 @@ namespace libp2p::protocol::kademlia {
...
+
     void truncate(size_t limit) {
-      if (size() > limit) {
-        erase(std::next(begin(), limit), end());
+      if (limit == 0) {
+        peers_.clear();
+      } else if (peers_.size() > limit) {
+        size_t n = peers_.size() - limit;
+        for (size_t i = 0; i < n; ++i) {
+          peers_.pop_back();
```
Comment (igor-egorov) [inline include/libp2p/protocol/kademlia/impl/peer_routing_table_impl.hpp:116]: (not required)
could be just ```peers_.resize(limit);```

File: include/libp2p/protocol/kademlia/impl/peer_routing_table_impl.hpp [hunk 2]
```diff
@@ -91,17 +142,25 @@ namespace libp2p::protocol::kademlia {
...
       Bucket b{};
-      // remove shifts all elements "to be removed" to the end, other elements
-      // preserve their relative order
-      auto new_end = std::remove_if(begin(), end(), [&](const auto &bpi) {
-        return bpi.node_id.commonPrefixLen(target) > commonLenPrefix;
-      });
 
-      b.assign(std::make_move_iterator(new_end),
-               std::make_move_iterator(end()));
+      std::list<BucketPeerInfo> newPeers;
```
Comment (igor-egorov) [inline include/libp2p/protocol/kademlia/impl/peer_routing_table_impl.hpp:146]: Google C++ style [guide](https://google.github.io/styleguide/cppguide.html#Variable_Names) prescribes to use snake case for names of that kind variables. (e.g. ```new_peers, common_len_prefix```).

**Review by igor-egorov — APPROVED** [review]

File: include/libp2p/protocol/kademlia/impl/peer_routing_table_impl.hpp (hunk 1 above)
Comment (art-gor) [reply include/libp2p/protocol/kademlia/impl/peer_routing_table_impl.hpp:116]: there is std::list, maybe this is cheaper

File: src/multi/multihash.cpp [hunk 3]
```diff
@@ -37,15 +38,47 @@ OUTCOME_CPP_DEFINE_CATEGORY(libp2p::multi, Multihash::Error, e) {
...
+    appendVarint(bytes, type);
+    BOOST_ASSERT(h.size() <= std::numeric_limits<uint8_t>::max());
+    bytes.push_back(static_cast<uint8_t>(h.size()));
+    hash_offset = bytes.size();
+    bytes.insert(bytes.end(), h.begin(), h.end());
+    std_hash = boost::hash_range(bytes.begin(), bytes.end());
+  }
+
+  const Multihash::Data &Multihash::data() const {
+#if THROW_IF_MOVED_OBJECT_ACCESSED
```
Comment (xDimon) [inline src/multi/multihash.cpp:69]: Think about using NDEBUG option

**Review by xDimon — APPROVED** [review]

**Pull request closed without merging**